ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommittingSpec.scala

```diff
@@ -0,0 +1,214 @@
+  private val metrics: Metrics = new Metrics(new MetricRegistry)
+  private val keyValueSubmission = new KeyValueSubmission(metrics)
+
+  private val alice = Party.assertFromString("Alice")
+  private val commandId = CommandId.assertFromString("cmdid")
+
+  private def toSubmission(tx: SubmittedTransaction): DamlSubmission = {
+    val timestamp = Time.Timestamp.assertFromInstant(Instant.EPOCH)
+    val meta = TransactionMeta(
+      ledgerEffectiveTime = timestamp,
+      workflowId = None,
+      submissionTime = timestamp,
+      submissionSeed = crypto.Hash.hashPrivateKey(this.getClass.getName),
+      optUsedPackages = Some(Set.empty),
+      optNodeSeeds = None,
+      optByKeyNodes = None,
+    )
+    val submitterInfo = SubmitterInfo(
+      actAs = List(alice),
+      applicationId = ApplicationId.assertFromString("appid"),
+      commandId = commandId,
+      deduplicateUntil = Instant.EPOCH,
+    )
+    keyValueSubmission.transactionToSubmission(
+      submitterInfo = submitterInfo,
+      meta = meta,
+      tx = tx,
+    )
+  }
+
+  def toSubmission(builder: TransactionBuilder): DamlSubmission =
+    this.toSubmission(builder.buildSubmitted())
+
+  val dedupKey = DamlStateKey.newBuilder
+    .setCommandDedup(
+      DamlCommandDedupKey.newBuilder.addSubmitters(alice).setCommandId(commandId).build
+    )
+    .build
+
+  private val keyValue = Value.ValueUnit
+  private val templateId = Identifier.assertFromString("pkg:M:T")
+
+  private def create(builder: TransactionBuilder, id: String, hasKey: Boolean = false) =
+    builder.create(
+      id = id,
+      template = templateId.toString,
+      argument = Value.ValueRecord(None, ImmArray.empty),
+      signatories = Seq(),
+      observers = Seq(),
+      key = if (hasKey) Some(keyValue) else None,
+    )
+
+  private def exercise(
+      builder: TransactionBuilder,
+      id: String,
+      hasKey: Boolean = false,
+      consuming: Boolean = true,
+  ) =
+    builder.exercise(
+      contract = create(builder, id, hasKey),
+      choice = "Choice",
+      argument = Value.ValueRecord(None, ImmArray.empty),
+      actingParties = Set(),
+      consuming = consuming,
+      result = Some(Value.ValueUnit),
+    )
+
+  private def fetch(builder: TransactionBuilder, id: String, byKey: Boolean) =
+    builder.fetch(contract = create(builder, id, hasKey = true), byKey = byKey)
+
+  private def lookup(builder: TransactionBuilder, id: String, found: Boolean) =
+    builder.lookupByKey(contract = create(builder, id, hasKey = true), found = found)
+
+  import Conversions.{contractIdToStateKey, contractKeyToStateKey}
+
+  "submissionOutputs" should {
+
+    "return a single output for a create without a key" in {
+      val builder = TransactionBuilder()
+      val createNode = create(builder, "#1")
+      builder.add(createNode)
+      val key = contractIdToStateKey(createNode.coid)
+      KeyValueCommitting.submissionOutputs(toSubmission(builder)) shouldBe Set(
```

Comment:
For easier readability of tests, we prefer to structure test cases in kvutils using the [AAA pattern](https://freecontent.manning.com/making-better-unit-tests-part-1-the-aaa-pattern/), a.k.a., arrange-act-assert.

Reply:
77b601294ff2d44322eb839c481ea50bd0ce8099 if I understand your suggestion correctly

Reply:
I was thinking of separating the arrange part from the rest by a newline but I'm fine with merging this as-is.